source/common/stats/symbol_table_impl.cc

```diff
@@ -75,11 +174,87 @@ Symbol SymbolTableImpl::toSymbol(absl::string_view sv) {
   return result;
 }
 
-absl::string_view SymbolTableImpl::fromSymbol(const Symbol symbol) const {
+absl::string_view SymbolTable::fromSymbol(const Symbol symbol) const
+    EXCLUSIVE_LOCKS_REQUIRED(lock_) {
   auto search = decode_map_.find(symbol);
   ASSERT(search != decode_map_.end());
   return search->second;
 }
 
+void SymbolTable::newSymbol() EXCLUSIVE_LOCKS_REQUIRED(lock_) {
+  if (pool_.empty()) {
+    next_symbol_ = ++monotonic_counter_;
+  } else {
+    next_symbol_ = pool_.top();
+    pool_.pop();
+  }
+  // This should catch integer overflow for the new symbol.
+  ASSERT(monotonic_counter_ != 0);
+}
+
+bool SymbolTable::lessThan(const StatName& a, const StatName& b) const {
+  // Constructing two temp vectors during lessThan is not strictly necessary.
+  // If this becomes a performance bottleneck (e.g. during sorting), we could
+  // provide an iterator-like interface for incrementally decoding the symbols
+  // without allocating memory.
+  SymbolVec av = SymbolEncoding::decodeSymbols(a.data(), a.numBytes());
+  SymbolVec bv = SymbolEncoding::decodeSymbols(b.data(), b.numBytes());
+  for (size_t i = 0, n = std::min(av.size(), bv.size()); i < n; ++i) {
+    if (av[i] != bv[i]) {
+      Thread::LockGuard lock(lock_);
+      return fromSymbol(av[i]) < fromSymbol(bv[i]);
+    }
+  }
+  return av.size() < bv.size();
+}
+
+StatNameStorage::StatNameStorage(absl::string_view name, SymbolTable& table) {
+  SymbolEncoding encoding = table.encode(name);
+  bytes_.reset(new uint8_t[encoding.bytesRequired()]);
+  encoding.moveToStorage(bytes_.get());
+}
+
+StatNameStorage::~StatNameStorage() {
+  // StatNameStorage is not fully RAII: you must call free(SymbolTable&) to
+  // decrement the reference counts held by the SymbolTable on behalf of
+  // this StatName.
+  ASSERT(bytes_.get() == nullptr);
+}
+
+void StatNameStorage::free(SymbolTable& table) {
+  table.free(statName());
+  bytes_.reset();
+}
+
+StatNameJoiner::StatNameJoiner(StatName a, StatName b) {
+  size_t a_size = a.numBytes();
+  size_t b_size = b.numBytes();
+  uint8_t* p = alloc(a_size + b_size);
+  memcpy(p, a.data(), a_size);
+  memcpy(p + a_size, b.data(), b_size);
+}
+
+StatNameJoiner::StatNameJoiner(const std::vector<StatName>& stat_names) {
+  size_t num_bytes = 0;
+  for (StatName stat_name : stat_names) {
+    num_bytes += stat_name.numBytes();
+  }
+  uint8_t* p = alloc(num_bytes);
+  for (StatName stat_name : stat_names) {
+    num_bytes = stat_name.numBytes();
+    ;
+    memcpy(p, stat_name.data(), num_bytes);
+    p += num_bytes;
+  }
+}
+
+uint8_t* StatNameJoiner::alloc(size_t num_bytes) {
+  bytes_.reset(new uint8_t[num_bytes + 2]);
+  uint8_t* p = bytes_.get();
+  *p++ = num_bytes & 0xff;
```

Comment:
uint16_t size assign?

Reply:
factored out length-assigner and commented there why this can't be done via uint16_t.